### DISCUSSION TITLE
Writing a custom download strategy from scratch?

### DISCUSSION THREAD

**Discussion by lmartinez-mirror** [Homebrew/discussions#574] [Tap maintenance and Homebrew development] [answered]
Hi, I'm looking into creating a cask for a program that only distributes binaries through Mega.nz. I thought about processing the package elsewhere in a build step, but I figured it'd be best to use a custom download strategy. My only question is, will I need to write it from scratch, and if so, are there any examples I can follow? The only requirement is `core/megatools`.

**Comment by jonchang** [comment, marked as answer]
Check out https://rubydoc.brew.sh/AbstractFileDownloadStrategy.html and its subclasses. Mostly what you'd need to do is properly implement `fetch` by using `cached_location`, `symlink_location`, and `temporary_path` correctly. Note that this would likely not be accepted into Homebrew itself, so you'd have to host this cask or formula in your own tap: https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap

**Reply by lmartinez-mirror** [reply]
Thank you so much! Does this implementation need to be in the same cask file or separate?

**Reply by gromgit** [reply]
It can be a separate library file. See my FUSE tap for how I accomplished a similar task (new requirement): https://github.com/gromgit/homebrew-fuse

**Reply by jonchang** [reply]
It can be in the same file or a separate file, but if it's a separate file you'll need to include the other file with something like `require "../lib/megatools_download_strategy.rb"`
